include/etcpal/cpp/inet.h

```diff
@@ -0,0 +1,658 @@
+/******************************************************************************
+ * Copyright 2019 ETC Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************
+ * This file is a part of EtcPal. For more information, go to:
+ * https://github.com/ETCLabs/EtcPal
+ ******************************************************************************/
+
+/// \file etcpal/cpp/inet.h
+/// \brief C++ wrapper and utilities for etcpal/inet.h
+
+#ifndef ETCPAL_CPP_INET_H_
+#define ETCPAL_CPP_INET_H_
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <iterator>
+#include <string>
+#include "etcpal/inet.h"
+
+namespace etcpal
+{
+/// \defgroup etcpal_cpp_inet etcpal_cpp_inet
+/// \ingroup etcpal_cpp
+/// \brief C++ utilities for the \ref etcpal_inet module.
+
+/// \ingroup etcpal_cpp_inet
+/// \brief A wrapper class for the EtcPal IP address type.
+///
+/// Provides C++ syntactic sugar for working with IP addresses.
+class IpAddr
+{
+public:
+  IpAddr() noexcept;
+  // Note: this constructor is not explicit by design, to allow implicit conversions e.g.
+  //   etcpal::IpAddr ip = etcpal_c_function_that_returns_ip();
+  IpAddr(const EtcPalIpAddr& c_ip) noexcept;
+  IpAddr& operator=(const EtcPalIpAddr& c_ip) noexcept;
+
+  IpAddr(uint32_t v4_data) noexcept;
+  IpAddr(const uint8_t* v6_data) noexcept;
+  IpAddr(const uint8_t* v6_data, unsigned long scope_id) noexcept;
+
+  const EtcPalIpAddr& get() const noexcept;
+  std::string ToString() const;
+  uint32_t v4_data() const noexcept;
+  const uint8_t* v6_data() const noexcept;
+  unsigned long scope_id() const noexcept;
+
+  bool IsValid() const noexcept;
+  etcpal_iptype_t type() const noexcept;
+  bool IsV4() const noexcept;
+  bool IsV6() const noexcept;
+  bool IsLinkLocal() const noexcept;
+  bool IsLoopback() const noexcept;
+  bool IsMulticast() const noexcept;
+  bool IsWildcard() const noexcept;
+
+  void SetAddress(uint32_t v4_data) noexcept;
+  void SetAddress(const uint8_t* v6_data) noexcept;
```

Comment:
Should we have a std::array<uint8_t, 16> version of this (and the getter)? (And similar thought for MAC addresses and the corresponding constructor of SockAddr.)

Reply:
I've been going back and forth on this. Really the best thing here would be to replace the `const uint8_t*` overload with a `std::span<uint8_t, 16>` which adds bounds safety without enforcing how/where the data is held before you call the function, but sigh... it's not available yet.

Adding an array overload seems a bit pointless when you already have the "less safe" `const uint8_t*` overload that you could just call with `my_array.data()`. But on the other hand, it's relatively harmless to just add the overload. Idk...

Reply:
Hmm. I was definitely thinking of it as especially being useful as the return type on the getter, but it's owning which means it would force a copy on return. 

I'm not sure what the best answer is here. I do agree that std::span feels most correct were we to have it though.

Reply:
I am just going to add the array overloads, it'll be useful in the case that someone wants to copy data out of these types, and if they don't want to copy they can use the pointer overloads.

Reply:
We can re-address this when span is available and in the tools we need to support.

Reply:
I experimented with this and ended up removing everything but the array getters (which @jtveite said would be the most useful, and I agree). Array constructors caused ambiguous overload problems, and the array setters made the interface feel too bloated. We can always add them if we need them.